app/services/discord/sync_profile_service.rb

```diff
@@ -1,42 +1,147 @@
-        configuration.default_role_ids
-      ].flatten.compact.uniq
-
-      return if role_ids.empty?
-
-      Discordrb::API::Server.update_member(
-        "Bot #{configuration.bot_token}",
-        configuration.server_id,
-        @user.discord_user_id,
-        roles: role_ids,
-        nick: nick_name
-      )
+      return false unless sync_ready?
+
+      rest_client =
+        Discordrb::API::Server.update_member(
+          "Bot #{configuration.bot_token}",
+          configuration.server_id,
+          user.discord_user_id,
+          roles: all_discord_role_ids,
+          nick: nick_name
+        )
+
+      if rest_client.code == 200
+        sync_and_cache_roles(rest_client)
+        true
+      else
+        error(t("error_while_syncing", { error_msg: @error_msg }))
+      end
     rescue Discordrb::Errors::UnknownMember
-      Rails.logger.error "Unknown member #{@user.discord_user_id}"
+      message =
+        t("unknown_member", variables: { member_id: user.discord_user_id })
+      Rails.logger.error(message)
       @user.update!(discord_user_id: nil)
+
+      error(message)
     rescue Discordrb::Errors::NoPermission
-      Rails.logger.error "No permission to update member #{@user.discord_user_id}"
+      message =
+        t("no_permission", variables: { member_id: user.discord_user_id })
+      Rails.logger.error(message)
+
+      error(message)
     rescue RestClient::BadRequest => e
-      Rails.logger.error "Bad request with discord_user_id: #{@user.discord_user_id}; #{e.response.body}"
+      message =
+        t(
+          "bad_request",
+          variables: {
+            member_id: user.discord_user_id,
+            error: e.response.body
+          }
+        )
+      Rails.logger.error(message)
+
+      error(message)
+    end
+
+    def sync_ready?
+      user.discord_user_id.present? && configuration.configured?
+    end
+
+    private
+
+    attr_reader :user, :additional_discord_role_ids
+
+    def t(key, variables = {})
+      I18n.t("services.discord.sync_profile_service.#{key}", **variables)
+    end
+
+    def sync_and_cache_roles(rest_client)
+      response_role_ids = JSON.parse(rest_client.body).dig("roles")
+
+      additional_synced_role_ids = additional_role_ids & response_role_ids
+
+      @warning_msg = t("sync_warning") if (
+        additional_role_ids - additional_synced_role_ids
+      ).present?
+
+      school
+        .discord_roles
+        .where(discord_id: additional_synced_role_ids)
+        .each do |role|
+          AdditionalUserDiscordRole.where(
+            user_id: user.id,
+            discord_role_id: role.id
+          ).first_or_create!
+        end
+
+      deleted_discord_role_ids =
+        user
+          .discord_roles
+          .where.not(discord_id: additional_synced_role_ids)
+          .pluck(:id)
+
+      AdditionalUserDiscordRole
+        .where(discord_role_id: deleted_discord_role_ids)
+        .where(user_id: user.id)
+        .delete_all
```

Comment:
`AddtiontionalUserDiscordRole` is a join table, storing all the roles relationships that a user is assigned to, which means there will be multiple entries for a user who has multiple discord roles assigned. Since discord admin can remove roles from a user over on discord, we would want to be in sync and destroy the mappings when admin removes roles from a user.